### PR TITLE
Support for ros launch

### PR DESCRIPTION
**Pull request by adityapande-1995**
This PR aims to add support for ros launch using a new bazel rule. It is based off of the ros_py_binary rule.

This is a bit tricky as roslaunch has the power to execute arbitrary nodes, executables and a whole bunch of actions that rely on paths and environment being set accordingly.

### Design
I'm piggybacking off the ``ros_py_binary()`` rule here. That rule makes sure my environment is set up properly (rmw, index paths, etc) so that a ros node can be run in a python process. If I run a subprocess here, using the ``ros2`` CLI, I inherit the environment for free, and don't need to worry about separate yaml/xml/python parsing.

### Problems and limitations
EDIT(eric): See #344

###  Example

See README updates in the PR

cc @tervay-bdai  

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake-ros/298)

### CODE REVIEW

**Comment by adityapande-1995** [comment]
I'm unable to use pip_install() rule to install a local python package (i.e. roslaunch_utils in this case) using pip. Is there a suggestion on if this is possible ?

**Comment by adityapande-1995** [comment]
Update : ``pip_install()`` is the wrong approach here, there is no way to use a local package using that rule. Open for review !

File: bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    # Actual launch files should be able to import the custom action now.
+    env = {**os.environ, 'PYTHONPATH': os.getcwd() + '/external/bazel_ros2_rules/ros2/tools:'
+           + os.environ['PYTHONPATH']}
+
+    launch_file_name = sys.argv[1]
+
+    roslaunch_cli = "./external/ros2/ros2"
+    action = "launch"
+    # TODO : Is there a better way to locate the launch file exactly ?
+    launch_file = find_rel_path(launch_file_name, os.getcwd())
```
Comment (adeeb10abbas) [inline bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py:49]: Is is possible to use custom bazel's python toolchain to do this instead of this?

Comment (EricCousineau-TRI) [reply bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py:49]: IIRC, you want to try and just use `env` vars (a la `source /opt/ros/humble/setup.bash`) via wrappings, like `rules_conda`.
That's a steep departure from current approach, so I'd deem it out of scope for this PR.

Feel free to file issue tho

**Comment by EricCousineau-TRI** [comment]
@IanTheEngineer Do you have time to review / land this PR on Mon / Tues?
If not, I can help to do so.

File: bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py (hunk 1 above)
Comment (adeeb10abbas) [reply bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py:49]: Gotcha thanks!

**Comment by ali-bdai** [comment]
@EricCousineau-TRI Let me know if there's anything I can do to expedite this. Thanks!

**Comment by adityapande-1995** [comment]
Rebased on main, open for review.

File: bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    # Required to get xml launch working, as we need to register an entry
+    # point for launch extensions.
+    os.system("pip install ../bazel_ros2_rules/ros2/tools/roslaunch_util >/dev/null 2>&1")
+    # Actual launch files should be able to import the custom action now.
+    env = {**os.environ, 'PYTHONPATH': os.getcwd() + '/external/bazel_ros2_rules/ros2/tools:'
+           + os.environ['PYTHONPATH']}
+
+    launch_file_name = sys.argv[1]
+
+    roslaunch_cli = "./external/ros2/ros2"
```
Comment (adityapande-1995) [inline bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py:46]: Can use bazel runfiles here, to extract the path.

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> Reviewed 5 of 10 files at r1, 1 of 1 files at r2, 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-ros/298)* status: 7 of 10 files reviewed, 7 unresolved discussions (waiting on @adeeb10abbas and @adityapande-1995)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#-NuAYqhd3p5EwZsEGRGX:-NuAYqhd3p5EwZsEGRGY:bpxihaf) (no related file):*
There is no documentation in this PR as to how to confirm that Bazel-ified launching works.

Can you take your overview comments, and place them in code in this PR?

___
___
*[`bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py` line 39 at r4](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#-NuAY11g2gCh76uu1xx-:-NuAY11g2gCh76uu1xx0:b-tr64ot) ([raw file](https://github.com/robotlocomotion/drake-ros/blob/a87e1fc3483134005a6777067074bc848202c7f2/bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py#L39)):*
> ```Python
>     # Required to get xml launch working, as we need to register an entry
>     # point for launch extensions.
>     os.system("pip install ../bazel_ros2_rules/ros2/tools/roslaunch_util >/dev/null 2>&1")
> ```

`pip install` should not be used.

Why is it needed here? What fails if you do not do this?
___
*[`bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py` line 46 at r4](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#gh-1541968867:-NuAY7xB8DEYx-p1CfuH:b-thz8fz) ([raw file](https://github.com/robotlocomotion/drake-ros/blob/a87e1fc3483134005a6777067074bc848202c7f2/bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py#L46)):*
<details><summary><i>Previously, adityapande-1995 (Aditya Pande) wrote…</i></summary><blockquote>

Can use bazel runfiles here, to extract the path.
</blockquote></details>

Yes, should certainly use runfiles.
___
*[`bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py` line 49 at r4](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#-NuAYBKBAARI6iPyoefp:-NuAYBKBAARI6iPyoefq:blmhx1g) ([raw file](https://github.com/robotlocomotion/drake-ros/blob/a87e1fc3483134005a6777067074bc848202c7f2/bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py#L49)):*
> ```Python
>     action = "launch"
>     # TODO : Is there a better way to locate the launch file exactly ?
>     launch_file = find_rel_path(launch_file_name, os.getcwd())
> ```

Can you reference what test might fail if you don't have this search functionality?

What is the normal mechanism by which `ros2 launch` works when locating a binary? (and by that token, `ros2 run`)
Does it just find whichever binary first?

(Can you cite the code where this happens?)
___
*[`ros2_example_bazel_installed/ros2_example_apps/BUILD.bazel` line 243 at r4](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#-NuAYeDGBRx3p8_W0fUk:-NuAYeDGBRx3p8_W0fUl:bq3cve7) ([raw file](https://github.com/robotlocomotion/drake-ros/blob/a87e1fc3483134005a6777067074bc848202c7f2/ros2_example_bazel_installed/ros2_example_apps/BUILD.bazel#L243)):*
> ```Starlark
>         ":eg_listener",
>     ],
> )
> ```

This should have a test showing that our Bazel spelling of `ros2 launch` works.
___
*[`ros2_example_bazel_installed/ros2_example_apps/BUILD.bazel` line 243 at r4](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#-NuAZ4LS6cYNkObKo8fQ:-NuAZ4LS6cYNkObKo8fR:b-6mzt2y) ([raw file](https://github.com/robotlocomotion/drake-ros/blob/a87e1fc3483134005a6777067074bc848202c7f2/ros2_example_bazel_installed/ros2_example_apps/BUILD.bazel#L243)):*
> ```Starlark
>         ":eg_listener",
>     ],
> )
> ```

nit Ideally, with our own provided `AMENT_PREFIX_PATH`, we should be able to call `ros2 launch` itself on our own targets.

Consider adding a test to this extent, or a TODO to that effect.

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-ros/298)* status: 7 of 10 files reviewed, 7 unresolved discussions (waiting on @adeeb10abbas and @adityapande-1995)

___
*[`bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py` line 39 at r4](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#-NuAY11g2gCh76uu1xx-:-NviX1qE9eFY37Ep3B4g:bty4iby) ([raw file](https://github.com/robotlocomotion/drake-ros/blob/a87e1fc3483134005a6777067074bc848202c7f2/bazel_ros2_rules/ros2/tools/roslaunch_util/roslaunch_util.py#L39)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

`pip install` should not be used.

Why is it needed here? What fails if you do not do this?
</blockquote></details>

Yeah, this def. shouldn't be used. It polluted my system, s.t. `ros2 launch` will try to pull in`ros2launch_util` even outside of Bazel.

I had to fix my system by deleting the following:
```
rm -rf ~/.local/lib/python3.10/site-packages/roslaunch_util*
```

**Comment by EricCousineau-TRI** [comment]
@ali-bdai can you review this and see if it meets your needs?

If you need something more, would you be up for taking over this PR?

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-ros/298)* status: 0 of 19 files reviewed, all discussions resolved (waiting on @adeeb10abbas)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#-NuAYqhd3p5EwZsEGRGX:-Nvilw6t4xRJXqizVzeb:b-896fix) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

There is no documentation in this PR as to how to confirm that Bazel-ified launching works.

Can you take your overview comments, and place them in code in this PR?
</blockquote></details>

Done.

___

**Review by adeeb10abbas — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#-:-NvlGNbIAszlJxhFGNny:bnfp4nl)* status: 0 of 19 files reviewed, all discussions resolved (waiting on @EricCousineau-TRI)

**Review by IanTheEngineer — APPROVED** [review]
:LGTM: as well. I appreciate the iterations you've gone through here, Aditya and your most recent updates Eric. This all looks reasonable to me. I created an issue to track the outlined limitations of this feature:

https://github.com/RobotLocomotion/drake-ros/issues/344

@EricCousineau-TRI and @ali-bdai , please feel free to add your thoughts there for future improvements.

> Reviewed 9 of 18 files at r5, 5 of 8 files at r7, 3 of 3 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-ros/298#-:-NvlIJve0v1DXchCEp_i:bwbwxrl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @adityapande-1995)

**Review by EricCousineau-TRI — APPROVED** [review]
> Reviewed 9 of 18 files at r5, 5 of 8 files at r7, 3 of 3 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/drake-ros/298)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @adityapande-1995)

**Pull request merged**